File: ios/TodayFlow/TodayFlow/Design/TodayRitualEntryIllustration.swift
import UIKit

/// Иллюстрация входа в ритуал: файлы в бандле, папка `today-ritual-entry/`
/// (см. `Resources/today-ritual-entry/ASSET_SPEC.txt`).
enum TodayRitualEntryIllustration {
    static func timeSlot(for date: Date = Date(), calendar: Calendar = .current) -> String {
        let h = calendar.component(.hour, from: date)
        if (5 ..< 12).contains(h) { return "morning" }
        if (12 ..< 17).contains(h) { return "day" }
        return "evening"
    }

    static func energyBand(score: Int) -> String {
        let s = min(100, max(0, score))
        if s < 40 { return "soft" }
        if s < 70 { return "balanced" }
        return "vivid"
    }

    static func candidateFileNames(dateISO: String, energyScore: Int, now: Date = Date()) -> [String] {
        let slot = timeSlot(for: now)
        let band = energyBand(score: energyScore)
        let bases = [
            "\(dateISO)-\(slot)-\(band)",
            "\(dateISO)-\(slot)",
            dateISO,
            "default-\(slot)-\(band)",
            "default-\(slot)",
            "default",
        ]
        let exts = ["webp", "png", "jpg", "jpeg"]
        var out: [String] = []
        for b in bases {
            for e in exts {
                out.append("\(b).\(e)")
            }
        }
        return out
    }

    static func loadUIImage(dateISO: String, energyScore: Int) -> UIImage? {
        let sub = "today-ritual-entry"
        for name in candidateFileNames(dateISO: dateISO, energyScore: energyScore) {
            let ns = name as NSString
            let base = ns.deletingPathExtension
            let ext = ns.pathExtension
            if let url = Bundle.main.url(forResource: base, withExtension: ext, subdirectory: sub),
               let img = UIImage(contentsOfFile: url.path) {
                return img
            }
        }
        return nil
    }
}

File: frontend/src/lib/todayRitualEntryIllustration.ts
/**
 * Иллюстрация входа в ритуал «Твой день»: публичные файлы в
 * `/images/today-ritual-entry/` (см. ASSET_SPEC.txt в той папке).
 */

export type RitualEntryTimeSlot = "morning" | "day" | "evening";

export type RitualEntryEnergyBand = "soft" | "balanced" | "vivid";

const ENTRY_BASE = "/images/today-ritual-entry";

/** Локальные часы устройства: утро / день / вечер. */
export function ritualEntryTimeSlot(now: Date = new Date()): RitualEntryTimeSlot {
  const h = now.getHours();
  if (h >= 5 && h < 12) return "morning";
  if (h >= 12 && h < 17) return "day";
  return "evening";
}

/** Грубый «тон дня» по шкале энергии 0–100 (как в герое Today). */
export function ritualEntryEnergyBand(energyScore: number): RitualEntryEnergyBand {
  const s = Math.max(0, Math.min(100, Math.round(energyScore)));
  if (s < 40) return "soft";
  if (s < 70) return "balanced";
  return "vivid";
}

function basesFor(dateISO: string, slot: RitualEntryTimeSlot, band: RitualEntryEnergyBand): string[] {
  return [
    `${dateISO}-${slot}-${band}`,
    `${dateISO}-${slot}`,
    dateISO,
    `default-${slot}-${band}`,
    `default-${slot}`,
    "default",
  ];
}

const EXT_ORDER = [".webp", ".png", ".jpg", ".jpeg"] as const;

/** Полные пути для `<img src>` — по порядку до первого успешно загруженного. */
/** Строка над заголовком: дата · Утро|День|Вечер (локальные часы `now`). */
export function ritualEntryEyebrowLine(displayDate: string, now: Date): string {
  const slot = ritualEntryTimeSlot(now);
  const word = slot === "morning" ? "Утро" : slot === "day" ? "День" : "Вечер";
  return `${displayDate} · ${word}`;
}

export function ritualEntryImagePublicPaths(
  dateISO: string,
  energyScore: number,
  now: Date = new Date(),
): string[] {
  const slot = ritualEntryTimeSlot(now);
  const band = ritualEntryEnergyBand(energyScore);
  const out: string[] = [];
  for (const base of basesFor(dateISO, slot, band)) {
    for (const ext of EXT_ORDER) {
      out.push(`${ENTRY_BASE}/${base}${ext}`);
    }
  }
  return out;
}
